AsyncDisplayKit/ASDisplayNode.mm

```diff
@@ -2060,78 +2179,53 @@ - (void)layout
-  for (ASLayout *subnodeLayout in _layout.sublayouts) {
-    if (![[self class] usesImplicitHierarchyManagement]) {
-      ASDisplayNodeAssert([_subnodes containsObject:subnodeLayout.layoutableObject], @"Sublayouts must only contain subnodes' layout.  self = %@, subnodes = %@", self, _subnodes);
-    }
-    CGPoint adjustedOrigin = subnodeLayout.position;
-    if (isfinite(adjustedOrigin.x) == NO) {
-      ASDisplayNodeAssert(0, @"subnodeLayout has an invalid position");
-      adjustedOrigin.x = 0;
-    }
-    if (isfinite(adjustedOrigin.y) == NO) {
-      ASDisplayNodeAssert(0, @"subnodeLayout has an invalid position");
-      adjustedOrigin.y = 0;
-    }
-    subnodeFrame.origin = adjustedOrigin;
-    
-    CGSize adjustedSize = subnodeLayout.size;
-    if (isfinite(adjustedSize.width) == NO) {
-      ASDisplayNodeAssert(0, @"subnodeLayout has an invalid size");
-      adjustedSize.width = 0;
-    }
-    if (isfinite(adjustedSize.height) == NO) {
-      ASDisplayNodeAssert(0, @"subnodeLayout has an invalid position");
-      adjustedSize.height = 0;
-    }
-    subnodeFrame.size = adjustedSize;
-    
-    subnode = ((ASDisplayNode *)subnodeLayout.layoutableObject);
-    [subnode setFrame:subnodeFrame];
-  }
   
   if ([[self class] usesImplicitHierarchyManagement]) {
-    for (_ASDisplayNodePosition *position in _deletedSubnodes) {
-      [self _implicitlyRemoveSubnode:position.node atIndex:position.index];
-    }
-    
-    for (_ASDisplayNodePosition *position in _insertedSubnodes) {
-      [self _implicitlyInsertSubnode:position.node atIndex:position.index];
+    [self __layoutSublayouts];
+  } else {
+    // Assume that _layout was flattened and is 1-level deep.
+    CGRect subnodeFrame = CGRectZero;
+    for (ASLayout *subnodeLayout in _layout.sublayouts) {
+      ASDisplayNodeAssert([_subnodes containsObject:subnodeLayout.layoutableObject], @"Sublayouts must only contain subnodes' layout. self = %@, subnodes = %@", self, _subnodes);
+      subnodeFrame = [self _adjustedFrameForLayout:subnodeLayout];
+      ((ASDisplayNode *)subnodeLayout.layoutableObject).frame = subnodeFrame;
     }
   }
 }
 
-- (void)_implicitlyInsertSubnode:(ASDisplayNode *)node atIndex:(NSUInteger)idx
+- (void)__layoutSublayouts
 {
-  ASDisplayNodeAssertThreadAffinity(self);
-
-  if (!_managedSubnodes) {
-    _managedSubnodes = [NSMutableArray array];
-  }
-
-  ASDisplayNodeAssert(idx <= [_managedSubnodes count], @"index needs to be in range of the current managed subnodes");
-  if (idx == [_managedSubnodes count]) {
-    [_managedSubnodes addObject:node];
-  } else {
-    [_managedSubnodes insertObject:node atIndex:idx];
+  for (ASLayout *subnodeLayout in _layout.sublayouts) {
+    ((ASDisplayNode *)subnodeLayout.layoutableObject).frame = [self _adjustedFrameForLayout:subnodeLayout];
   }
-  [self addSubnode:node];
 }
 
-- (void)_implicitlyRemoveSubnode:(ASDisplayNode *)node atIndex:(NSUInteger)idx
+- (CGRect)_adjustedFrameForLayout:(ASLayout *)layout
```

Comment:
Thanks for creating this method!  Should it be a C function for performance, if changing to ASDisplayNodeCAssert would remove the last dependency on self? (I may have missed one).  If it is only called in one place, then we could also inline it without code bloat.